### PR TITLE
Allow controlling the prefix added to repos/packages

### PR DESCRIPTION
**Pull request by person142**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Does not include precompiled binaries, eg. `.par` files. See [CONTRIBUTING.md](https://github.com/bazelbuild/rules_python/blob/master/CONTRIBUTING.md) for info
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature (please, look at the "Scope of the project" section in the README.md file)
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## What is the current behavior?
Issue Number:
- https://github.com/bazelbuild/rules_python/issues/414
- Closes https://github.com/bazelbuild/rules_python/issues/450

Currently the prefix added to the generated repos for `pip_parse` is hard-coded to `<pip-parse-name>_pypi__` and the prefix added to the generated packages for `pip_install` is hard-coded to `pypi__`.

## What is the new behavior?
Allow the user to control the prefix. You still need some prefix to avoid https://github.com/bazelbuild/rules_python/issues/414#issuecomment-789416360, but for `pip_parse` in particular you can make the external repository names more memorable:

```
$ cat requirements.txt
attrs==20.3.0
$ cat WORKSPACE
local_repository(
    name = "rules_python",
    path = "../rules_python",
)

load("@rules_python//python:pip.bzl", "pip_install", "pip_parse")

pip_parse(
    name = "deps",
    requirements_lock = "//:requirements.txt",
    repo_prefix="pypi_",
)

load("@deps//:requirements.bzl", "install_deps")
install_deps()
$ bazelisk build @pypi_attrs//:pkg
INFO: Analyzed target @pypi_attrs//:pkg (0 packages loaded, 0 targets configured).
INFO: Found 1 target...
Target @pypi_attrs//:pkg up-to-date (nothing to build)
INFO: Elapsed time: 0.095s, Critical Path: 0.00s
INFO: 1 process: 1 internal.
INFO: Build completed successfully, 1 total action
```

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

## Other information
None

### CODE REVIEW

File: python/pip.bzl [hunk 1]
```diff
@@ -45,26 +45,46 @@ def pip_install(requirements, name = "pip", **kwargs):
...
     pip_repository(
         name = name,
         requirements = requirements,
+        repo_prefix = repo_prefix,
         **kwargs
     )
 
-def pip_parse(requirements_lock, name = "pip_parsed_deps", **kwargs):
+def pip_parse(requirements_lock, name = "pip_parsed_deps", repo_prefix = None, **kwargs):
+    """Generate external repositories for each entry in a requirements lockfile.
```
Comment (person142) [inline python/pip.bzl:66]: Buildifier got unhappy that there was no docstring here.

File: python/pip.bzl [hunk 2]
```diff
@@ -45,26 +45,46 @@ def pip_install(requirements, name = "pip", **kwargs):
...
+      repo_prefix: prefix for the generated external repositories
+      **kwargs: keyword arguments passed directly to the `pip_repository` repository rule.
+
+    """
+
     # Just in case our dependencies weren't already fetched
     pip_install_dependencies()
 
+    if repo_prefix == None:
+        repo_prefix = "{name}_pypi__".format(name = name)
```
Comment (hrfuller) [inline python/pip.bzl:81]: Without the parent repo name in the repo_prefix these repositories won't be unique, and if users have multiple instances of pip_parse they could collide. I'm not sure I'm in favor of allowing that.

Comment (person142) [reply python/pip.bzl:81]: We could require a pattern like `<repo-name>_<pkg-name>`-my biggest objective is to have short, memorable names that are easy to refer to by label.

Comment (person142) [reply python/pip.bzl:81]: i.e. most of all I want to solve https://github.com/bazelbuild/rules_python/issues/414, which just so happens to be easier in the `pip_parse` case than the `pip_install` case...

**Comment by thekyz** [comment]
What's the status of this PR? In monorepos with a lot of packages controlling that prefix can be a life saver.

**Comment by thekyz** [comment]
@thundergolfer would you be willing to review this if I updated the PR? We currently have a bad alternative to this as a patch in our repo so I'd prefer to have it clean from mainstream.

**Comment by person142** [comment]
I can update this if desired/needed-just don't have the bandwidth right now to rally support for a particular solution.

**Comment by hrfuller** [comment]
@thekyz @person142 taking another look at this today. If we can resolve  the naming collision issue I'll be happy to ship.

File: python/pip.bzl (hunk 2 above)
Comment (hrfuller) [reply python/pip.bzl:81]: @person142 if we can make these repo names unique using `<repo-name>_<pkg-name>` Where <pkg-name> is the [PEP503 canonical package name](https://www.python.org/dev/peps/pep-0503/#normalized-names) I think thats a good soln.

Comment (person142) [reply python/pip.bzl:81]: 👍 Should have some time to make the change today.

**Comment by person142** [comment]
@hrfuller refactored so that the repos for `pip_parse` are now always `<name>_<package-name>`.

**Review by hrfuller — APPROVED** [review]
LGTM thanks @person142

**Pull request merged**